tools/fasm2frame.py

```diff
@@ -140,53 +206,60 @@ def frame_set(frame_addr, word_addr, bit_index):
         for coli in range(segj['frames']):
             frame_init(seg_baseaddr + coli)
 
-        # Now lets look up the bits we need frames for
-        segdb = get_database(segj['type'])
-
-        def clb2dbkey(tile, tilej, site, suffix, value):
-            db_k = '%s.%s.%s' % (tilej['type'], site, suffix)
-            return db_k
-
-        def int2dbkey(tile, tilej, site, suffix, value):
-            return '%s.%s.%s' % (tilej['type'], suffix, value)
-
-        tile2dbkey = {
-            'CLBLL_L': clb2dbkey,
-            'CLBLL_R': clb2dbkey,
-            'CLBLM_L': clb2dbkey,
-            'CLBLM_R': clb2dbkey,
-            'INT_L': int2dbkey,
-            'INT_R': int2dbkey,
-            'HCLK_L': int2dbkey,
-            'HCLK_R': int2dbkey,
-        }
-
-        f = tile2dbkey.get(tilej['type'], None)
-        if f is None:
-            raise Exception("Unhandled segment type %s" % tilej['type'])
-        db_k = f(tile, tilej, site, suffix, value)
-
-        try:
-            db_vals = segdb[db_k]
-        except KeyError:
-            raise Exception(
-                "Key %s (from line '%s') not found in segment DB %s" %
-                (db_k, l, segj['type']))
-
-        for val in db_vals:
-            # Default is 0. Skip explicit call outs
-            if val[0] == '!':
-                continue
-            # 28_05 => 28, 05
-            seg_word_column, word_bit_n = val.split('_')
-            seg_word_column, word_bit_n = int(seg_word_column), int(word_bit_n)
+        def update_segbit(seg_word_column, word_bit_n, isset):
+            '''Set  or clear a single bit in a segment at the given word column and word bit position'''
             # Now we have the word column and word bit index
             # Combine with the segments relative frame position to fully get the position
             frame_addr = seg_baseaddr + seg_word_column
             # 2 words per segment
             word_addr = seg_word_base + word_bit_n // 32
             bit_index = word_bit_n % 32
-            frame_set(frame_addr, word_addr, bit_index)
+            if isset:
+                frame_set(frame_addr, word_addr, bit_index)
+            else:
+                frame_clear(frame_addr, word_addr, bit_index)
+
+        # Now lets look up the bits we need frames for
+        segdb = get_database(segj['type'])
+
+        db_k = '%s.%s' % (tilej['type'], name)
+        try:
+            db_vals = segdb[db_k]
+        except KeyError:
+            raise FASMSyntaxError(
+                "Segment DB %s, key %s not found from line '%s'" %
+                (segj['type'], db_k, l))
+        '''
+        Creating the key depends on whether its an enumerated or a fixed value
+        If its enumerated, the value forms part of the key
+        We'd also like to ideally identify if there is a syntax error
+        Ultimately probably easier to separate out these two cases
+        '''
+        # A single bit value?
+        if None in db_vals:
```

Comment:
I'd rather see 1 used as the key in db_vals and this test changed to len(db_vals) == 1.

Reply:
Better yet, for single bit values, insert two keys for 0 and 1.  If no value is provided in the FASM line, assume they meant 1.  Then you can use the same code path for looking up values here.

Reply:
I don't like the first suggestion, but second sounds very reasonable to me. Changed in PR to that